Add Transaction Censorship Detection doc

jhaaaa opened the pull request:
- Adds a concept doc on transaction censorship detection and the transaction censorship detector.
- Adds a link to this concept doc to the Censorship-Resistant Transaction Processing section of the XRP Ledger Overview page.

jhaaaa committed:
first draft

jhaaaa commented:
@nbougalis - thank you very much for reviewing my very rough first draft and providing so many key details! I've incorporated those details to provide this PR. 

@mDuo13 - if you have time and interest, please feel free to provide feedback. Also, I'm wondering why other commits to the rippled_1.2.0 branch are showing up in this PR. 🤔

mDuo13 commented on content/concepts/introduction/xrp-ledger-overview.md:
```diff
@@ -30,6 +30,8 @@ This combines qualities of physical and centralized digital money. Prior to the
 
 The XRP Ledger's system of trusted validators uses a small amount of human interaction to achieve better distribution of authority than other decentralized systems. Fully-automated systems for reaching consensus from an unknown set of participants are vulnerable to concentrations of voting power. For example, Bitcoin mining is disproportionately concentrated in places with cheap electricity. As R ...
 
+For more information about the XRP Ledger's ability to detect censorship, see [Transaction Censorship Detection](transaction-censorship-detection.html). ***TODO: appropriate here?***
```
:+1:

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+# Transaction Censorship Detection
+
+The XRP Ledger is designed to be [censorship resistant](xrp-ledger-overview.html#censorship-resistant-transaction-processing). Any attempt to censor transactions requires coordinated action by a majority of the XRP Ledger network's validators.
```
I don't think it requires coordination by a majority of validators to _attempt_ to censor transactions. I mean, whether they succeed is a different matter. 😉

Especially if it's just delaying transactions to later ledger versions (which could be relevant when it comes to chasing high-value offers in the decentralized exchange, aka front-running), I would think that temporarily censoring certain transactions—just delaying them a ledger version or two—might be possible with a relatively-small number of validators. If transactions that are on the "avalanche" threshold of being included in a consensus round or not, it seems like a few validators could be the deciding vote causing those transactions to be delayed to the next ledger version.

I don't know whether this censorship detector would recognize that sort of behavior, but it is a malicious abuse of consensus that seems possible.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+
+While a `rippled` server is in sync with the network, the detector tracks all transactions that, in the view of the `rippled` server, should have been accepted in the last round of [consensus](intro-to-consensus.html) and included in the last validated ledger. The detector issues log messages of increasing severity for transactions that have not been included in a validated ledger after several r ...
+
+A `rippled` server is considered to be in sync if it meets the following criteria: ***TODO: Is this info useful or overkill?***
```
I think maybe overkill, especially since some of these points are prerequisites for others.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+- It has good connectivity.
+- It is tracking the current consensus round.
+
+At a high-level, here’s how the transaction censorship detector works:
```
I think breaking this into a separate sub-heading ("How it works") makes some sense.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+
+XRP Ledger provides an automated transaction censorship detector that is available on all `rippled` servers, enabling all participants to see if censorship is affecting the network.
+
+While a `rippled` server is in sync with the network, the detector tracks all transactions that, in the view of the `rippled` server, should have been accepted in the last round of [consensus](intro-to-consensus.html) and included in the last validated ledger. The detector issues log messages of increasing severity for transactions that have not been included in a validated ledger after several r ...
```
I'd like a sentence or even a section on false positives, innocent reasons that censorship transaction might trigger. (A likely one is a bug in relaying transactions to other servers in the network.)

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+
+This is an example warning message issued by the transaction censorship detector after transaction E08D6E9754025BA2534A78707605E0601F03ACE063687A0CA1BDDACFCD1698C7 remained in the tracker for 15 ledgers, from ledger 18851530 to ledger 18851545.
+
+        Potential Censorship: Eligible tx E08D6E9754025BA2534A78707605E0601F03ACE063687A0CA1BDDACFCD1698C7, which we are tracking since ledger 18851530 has not been included as of ledger 18851545.
```
I'm guessing these log messages probably start with a subsystem and warning level (such as `LedgerConsensus:ERR`), which we should include. We can omit the timestamp, though. (Note to self: should make the log messages in https://developers.ripple.com/understanding-log-messages.html more consistent.)

jhaaaa replied:
Thank you for this - confirmed with Nik that these values are `LedgerConsensus:WRN` and `LedgerConsensus:ERR`.

mDuo13 reviewed (approved):
Mostly looks good. Left a few suggestions and of course I assume you'll edit out the TODOs before merging.

One more suggestion: I think it's worth adding an entry in the "[Understanding Log Messages](https://developers.ripple.com/understanding-log-messages.html)" troubleshooting page to link back to this article.

As for why changes that should already be in the rippled_1.2.0 are showing up in this and other PRs, I'm not sure yet.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -0,0 +1,42 @@
+# Transaction Censorship Detection
+
```
Oh, add a `[New in: rippled 1.2.0][]` badge here

mDuo13 commented:
Re: the extra commits, [I figured it out](https://github.com/ripple/ripple-dev-portal/pull/502#issuecomment-458885887).

tl;dr rebase this on the latest `rippled_1.2.0` branch to fix it. (`git checkout` and `git pull` the latest `rippled_1.2.0` branch, then `git checkout` this branch again and, `git rebase` it on (the now-updated) `rippled_1.2.0` and finally force push this branch)

jhaaaa committed:
inc'd feedback

jhaaaa commented:
@mDuo13 - Thank you for the great review and suggestions. I've updated the content to address. I've also rebased based on your instructions - thank you! The Jenkins X is due to the broken link to `https://github.com/ripple/rippled/releases/tag/1.2.0`, which is pending. 

@nbougalis - thank you for providing additional info about potential false positives. If you have the time, please feel free to provide feedback.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -23,20 +20,37 @@ At a high-level, here’s how the transaction censorship detector works:
     For as long as the transaction remains in the tracker, the detector continues to issue a warning message in the log every 15 ledgers, for up to five warning messages. After the fifth warning message, the detector issues a final [error message](#example-error-message) in the log and then stops issuing warning and error messages.
 
-    If you see these messages in your `rippled` server log, you may want to investigate, off-ledger, why other `rippled` servers are failing to include the transaction.
+    If you see these messages in your `rippled` server log, first verify that the messages are not [false positives](#potential-false-positives). If they are not false positives, you may want to investigate, off-ledger, why other `rippled` servers are failing to include the transaction.
```
Nit: I would frame this slightly differently. Regardless of whether it's a false positive, you should investigate why other servers are failing to include the transaction, starting with the assumption that this is more likely to be a false positive (innocent bug) rather than malicious censorship.

mDuo13 commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
```diff
@@ -154,6 +154,10 @@ This is not strictly a problem, but if you want to acquire ledger history faster
+## Potential Censorship
+
+The transaction censorship detector issues log messages when it detects potential transaction censorship. For more information about the detector and these log messages, see [Transaction Censorship Detection](transaction-censorship-detection.html).
+
```
Please include an example of the log messages in question here, as demonstrated in the other examples on the page. The intent is for this page to be searchable by Ctrl-F if you have any log messages you don't understand.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -23,20 +20,37 @@ At a high-level, here’s how the transaction censorship detector works:
+
+- Your `rippled` server is out of sync with the network and has not yet realized it.
+
+- Your `rippled` server is being impacted by a class of bug that causes `rippled` servers to inconsistently relay transactions to other `rippled` servers in the network.
```
Even if _your_ `rippled` server isn't impacted by this class of bug, if enough other `rippled` servers in the network are affected by this bug, it could result in the same symptoms.

mDuo13 reviewed (approved):
Requested some fairly small changes, just to be very precise with the wording we use here.

jhaaaa committed:
inc'd feedback

jhaaaa commented:
@mDuo13 - thank you for your patience in providing these clarifications. 🙌  I believe I've addressed your latest feedback.

mDuo13 commented on content/concepts/the-rippled-server/transaction-censorship-detection.md:
```diff
@@ -50,7 +52,7 @@ Here are some scenarios that could cause the detector to issue false positive me
 - Your `rippled` server is out of sync with the network and has not yet realized it.
 
-- Your `rippled` server is being impacted by a class of bug that causes `rippled` servers to inconsistently relay transactions to other `rippled` servers in the network.
+- `rippled` servers in the network, including possibly your own server, are being impacted by a class of bug that causes `rippled` servers to inconsistently relay transactions to other `rippled` servers in the network.
```
Last nit: Maybe pitch the bug bounty again here? Or at least encourage people to report the bug if they're seeing this? Just to be clear that there are no currently-known bugs and this is _not_ expected behavior.

jhaaaa replied:
@mDuo13 - fantastic addition. Just added it... 🙌

mDuo13 reviewed (approved):
:+1:

jhaaaa committed:
bug bounty ref

jhaaaa commented:
The build failure is due to the broken link to `https://github.com/ripple/rippled/releases/tag/1.2.0`, which will be live when this doc is published.
